resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosCoarseGrainedSchedulerBackend.scala

```diff
@@ -791,7 +795,13 @@ private[spark] class MesosCoarseGrainedSchedulerBackend(
     // We don't truly know if we can fulfill the full amount of executors
     // since at coarse grain it depends on the amount of slaves available.
     logInfo("Capping the total amount of executors to " + requestedTotal)
+    val reviveNeeded = executorLimit < requestedTotal
     executorLimitOption = Some(requestedTotal)
+    if (reviveNeeded && schedulerDriver != null) {
```

Comment:
I'm wondering if this is safe to rely on `doRequestTotalExecutors` invocations to revive offers. In addition to this check,  I'd suggest running another check in a background thread and revive offers if the conditions apply.

Reply:
I think all cases are covered actually. There are two conditions when offers are suppressed:

1. `numExecutors >= executorLimit`

   - `numExecutors` decreases only when `executorTerminated()` is called, which happens only `statusUpdate()` right before reviving offers

   - `executorLimit` changes only in `doRequestTotalExecutors()`

2. `totalCoresAcquired >= maxCores`

   - Similarly to `numExecturos`, `totalCoresAcquired` decreases only in `statusUpdate()` when a task has finished, which is always followed by a revive.

   - `maxCores` is a pre-configured value and doesn't change

Reply:
See my comment above regarding `revive`s potentially being dropped. So we need to hook into some kind of periodic thread to check if there are executors/tasks to launch and do a revive to be safe.